pkg/delete/internal/bucket/delete.go

```diff
@@ -0,0 +1,128 @@
+/*
+ * @license
+ * Copyright 2023 Dynatrace LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bucket
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api/clients/buckets"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/idutils"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log/field"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/delete/pointer"
+	"golang.org/x/net/context"
+	"net/http"
+	"strings"
+)
+
+type Client interface {
+	Delete(ctx context.Context, id string) (buckets.Response, error)
+	List(ctx context.Context) (buckets.ListResponse, error)
+}
+
+func Delete(ctx context.Context, c Client, entries []pointer.DeletePointer) error {
+
+	logger := log.WithCtxFields(ctx).WithFields(field.Type("bucket"))
+	logger.Info(`Deleting %d config(s) of type "bucket"...`, len(entries))
+
+	deleteErrs := 0
+	for _, e := range entries {
+
+		logger := logger.WithFields(field.Coordinate(e.AsCoordinate()))
+
+		bucketName := idutils.GenerateBucketName(e.AsCoordinate())
+
+		logger.Debug("Deleting bucket: %s.", e, bucketName)
+		resp, err := c.Delete(ctx, bucketName)
+		if err != nil {
+			logger.WithFields(field.Error(err)).Error("Failed to delete Grail Bucket configuration: %v", e, bucketName, err)
+			deleteErrs++
```

Comment:
@warber also commented that on another PR 🤔 
In general, this is fine: I'm using the fact that the new clients return an error OR a response. If there was an err return, there's no response that will be an APIError. 

But I can see how it looks odds - especially if @warber who introduced the pattern stumbles over it as well :D

I'll take a look at making this more explicitly count an error it err or APIError happened

Reply:
hmmm.. as nice as the very strict separation between these kinds of errors is, I fear it will introduce checking errors kind of "twice" in a row in a lot of situtations, which could indeed look a bit odd. hmm

Reply:
So I've tried to simplify this, but IMO failed/it's not much better, so I didn't commit it.

Generally I still think the separation/not assuming what should be an error kinda makes sense - but it might be that every user, not just monaco will always check for error & APIError...
Wrapping a helper function around that doesn't solve what might be the wrong design after all...

```go
func deleteBuckets(ctx context.Context, c bucketClient, entries []DeletePointer) []error {
	log.WithCtxFields(ctx).WithFields(field.Type("bucket")).Info("Deleting %d config(s) of type %q...", len(entries), "bucket")
	errors := make([]error, 0)
	for _, e := range entries {
		bucketName := idutils.GenerateBucketName(e.asCoordinate())
		resp, err := c.Delete(ctx, bucketName)
		if err, failed := deleteFailed(resp, err); failed {
			errors = append(errors, fmt.Errorf("could not delete Bucket Definition object %s with name %q: %w", e, bucketName, err))
		}
	}
	return errors
}

func deleteFailed(resp buckets.Response, err error) (error, bool) {
		if err != nil {
			return err, true
		}
		if err, ok := resp.AsAPIError(); ok && err.StatusCode != http.StatusNotFound {
			return err, true
		}
		return nil, false
	}
```

Reply:
meh, that doesn't solve anything...
I still like it and imo it just makes sense... e.g. when the API returns a 404, then the call **was successful, hence no error.
To hide that behavior, another type of client on another level which even hides the fact that we are speaking HTTP/Rest needs to be introduced which uses the API client(s) and gives you very specific errors that make sense. But I think we don't need them, at least not in the lib scope

Reply:
Agreed in internal discussion to: 

1. just introduce if-elseif to make this readable without knowing implementation details
2. use the fact that we have distinct errors to have clear user messages if a network issue occurred or the API rejected something. 

This will be addressed on this branch/PR after #1191 is merged and rebased - as that introduces the same patterns for Automation client